pymc_extras/statespace/models/DFM.py

```diff
@@ -0,0 +1,828 @@
+    a AR(2) structure on the factor and a AR(1) structure on the errors:
+
+    .. code:: python
+
+        import pymc_extras.statespace as pmss
+        import pymc as pm
+
+        # Create DFM Statespace Model
+        dfm_mod = pmss.BayesianDynamicFactor(
+                k_factors=1,                # Number of latent dynamic factors
+                factor_order=2,             # Number of lags for the latent factor process
+                endog_names=data.columns,   # Names of the observed time series (endogenous variables) (we could also use k_endog = len(data.columns))
+                error_order=1,              # Order of the autoregressive process for the observation noise (i.e., AR(q) error, here q=1)
+                error_var=False,            # If False, models errors as separate AR processes
+                error_cov_type="diagonal",  # Structure of the observation error covariance matrix: uncorrelated noise across series
+                measurement_error=True,     # Whether to include a measurement error term in the model
+                verbose=True
+            )
+
+        # Unpack coords
+        coords = dfm_mod.coords
+
+
+        with pm.Model(coords=coords) as pymc_mod:
+            # Priors for the initial state mean and covariance
+            x0 = pm.Normal("x0", dims=["state_dim"])
+            P0 = pm.Normal("P0", dims=["state_dim", "state_dim"])
+
+            # Factor loadings: shape (k_endog, k_factors)
+            factor_loadings = pm.Normal("factor_loadings", sigma=1, dims=["k_endog", "k_factors"])
+
+            # AR coefficients for factor dynamics: shape (k_factors, factor_order)
+            factor_ar = pm.Normal("factor_ar", sigma=1, dims=["k_factors", "factor_order"])
+
+            # AR coefficients for observation noise: shape (k_endog, error_order)
+            error_ar = pm.Normal("error_ar", sigma=1, dims=["k_endog", "error_order"])
+
+            # Std devs for observation noise: shape (k_endog,)
+            error_sigma = pm.HalfNormal("error_sigma", dims=["k_endog"])
+
+            # Observation noise covariance matrix
+            obs_sigma = pm.HalfNormal("sigma_obs", dims=["k_endog"])
+
+            # Build the symbolic graph and attach it to the model
+            dfm_mod.build_statespace_graph(data=data, mode="JAX")
+
+            # Sampling
+            idata = pm.sample(
+                draws=500,
+                chains=2,
+                nuts_sampler="nutpie",
+                nuts_sampler_kwargs={"backend": "jax", "gradient_backend": "jax"},
+            )
+
+    """
+
+    def __init__(
+        self,
+        k_factors: int,
+        factor_order: int,
+        k_endog: int | None = None,
+        endog_names: Sequence[str] | None = None,
+        k_exog: int | None = None,
+        exog_names: Sequence[str] | None = None,
+        shared_exog_states: bool | None = None,
+        exog_innovations: bool | None = None,
```

Comment:
This can have a default value of False, as can the other boolean arguments.

You only need to do the `type | None = None` thing if `type` is a mutable type, like a string or list. The issue is that these defaults can be mutated in-place, leading to unexpected side-effects. With non-mutable types like tuple, float, int, bool etc, you don't have to worry about this.

Here's an example of why it matters:

```py
def x(l: list[str] = ['a']):
    return l
res = x() 
res # ['a'], unsurprising
res.append('b') # Innocently keep working with the output
x() # ['a', 'b'], default argument modified in-place! 
```